power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/base_optimizer.hpp

```diff
@@ -0,0 +1,87 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+
+#include "../main_core/state.hpp"
+
+#include <concepts>
+
+namespace power_grid_model::optimizer {
+
+namespace detail {
+template <typename StateCalculator, typename State>
+concept state_calculator_c = main_core::main_model_state_c<State> &&
+                             std::invocable<std::remove_cvref_t<StateCalculator>, std::remove_cvref_t<State> const&>;
+
+template <typename StateCalculator, typename State_>
+    requires state_calculator_c<StateCalculator, State_>
+struct state_calculator_type {
+    using Calculator = StateCalculator;
+    using State = State_;
+    using result_type = std::invoke_result_t<std::remove_cvref_t<StateCalculator>, State_ const&>;
+};
+
+template <typename StateCalculator, typename State_>
+using state_calculator_result_t = typename state_calculator_type<StateCalculator, State_>::result_type;
+
+template <typename StateCalculator, typename State>
+concept steady_state_calculator_c =
+    steady_state_math_output_type<typename detail::state_calculator_result_t<StateCalculator, State>::value_type> &&
+    std::same_as<detail::state_calculator_result_t<StateCalculator, State>,
+                 std::vector<typename detail::state_calculator_result_t<StateCalculator, State>::value_type>>;
+
+template <typename StateCalculator, typename State_>
+    requires main_core::main_model_state_c<State_> &&
+             std::invocable<std::remove_cvref_t<StateCalculator>, State_ const&>
+class BaseOptimizer {
+  public:
+    using Calculator = StateCalculator;
+    using State = State_;
+    using ResultType = state_calculator_result_t<Calculator, State>;
+
+    BaseOptimizer() = default;
+    BaseOptimizer(BaseOptimizer const&) = delete;
+    BaseOptimizer(BaseOptimizer&&) noexcept = default;
+    BaseOptimizer& operator=(BaseOptimizer const&) = delete;
+    BaseOptimizer& operator=(BaseOptimizer&&) noexcept = default;
+    virtual ~BaseOptimizer() = default;
+
+    virtual auto optimize(State const& state) -> ResultType = 0;
```

Comment:
How would this interface facilitate the fact that it modifies states in the middle, i.e., `math_model_state` ?

Reply:
that goes via the `update` function that is provided